pkg/prometheus/statefulset.go

```diff
@@ -321,48 +322,29 @@ func makeStatefulSetSpec(
 		return nil, err
 	}
 
-	version, err := semver.ParseTolerant(promVersion)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to parse prometheus version")
-	}
-
-	if version.Major != 2 {
-		return nil, errors.Errorf("unsupported Prometheus major version %s", version)
+	if cg.version.Major != 2 {
+		return nil, errors.Errorf("unsupported Prometheus major version %s", cg.version)
 	}
 
 	promArgs := []monitoringv1.Argument{
 		{Name: "web.console.templates", Value: "/etc/prometheus/consoles"},
 		{Name: "web.console.libraries", Value: "/etc/prometheus/console_libraries"},
 	}
 
-	// TODO(simonpasquier): log a warning message if the Prometheus version
-	// doesn't support the flag (do it everywhere it needs to be, not only for
-	// this block).
-	retentionTimeFlag := monitoringv1.Argument{Name: "storage.tsdb.retention"}
-	if version.GTE(semver.MustParse("2.7.0")) {
-		retentionTimeFlag = monitoringv1.Argument{Name: "storage.tsdb.retention.time"}
-		if p.Spec.Retention == "" && p.Spec.RetentionSize == "" {
-			retentionTimeFlag.Value = defaultRetention
-			promArgs = append(promArgs, retentionTimeFlag)
-		} else {
-			if p.Spec.Retention != "" {
-				retentionTimeFlag.Value = string(p.Spec.Retention)
-				promArgs = append(promArgs, retentionTimeFlag)
-			}
-
-			if p.Spec.RetentionSize != "" {
-				retentionSizeFlag := monitoringv1.Argument{Name: "storage.tsdb.retention.size", Value: string(p.Spec.RetentionSize)}
-				promArgs = append(promArgs, retentionSizeFlag)
-			}
-		}
-	} else {
-		if p.Spec.Retention == "" {
-			retentionTimeFlag.Value = defaultRetention
-			promArgs = append(promArgs, retentionTimeFlag)
-		} else {
-			retentionTimeFlag.Value = string(p.Spec.Retention)
-			promArgs = append(promArgs, retentionTimeFlag)
-		}
+	retentionTimeFlag := monitoringv1.Argument{Name: "storage.tsdb.retention.time", Value: defaultRetention}
+	if cg.WithMaximumVersion("2.7.0").IsCompatible() {
+		retentionTimeFlag.Name = "storage.tsdb.retention"
+	}
+	if p.Spec.Retention == "" && (p.Spec.RetentionSize == "" || cg.version.LT(semver.MustParse("2.7.0"))) {
+		promArgs = append(promArgs, retentionTimeFlag)
+	}
+	if p.Spec.Retention != "" {
+		retentionTimeFlag.Value = string(p.Spec.Retention)
+		promArgs = append(promArgs, retentionTimeFlag)
+	}
```

Comment:
(nit) maybe a bit easier to read?
```suggestion
	var (
		retentionTimeFlagName = "storage.tsdb.retention.time"
		retentionTimeFlagValue = p.Spec.Retention
	)
	if cg.WithMaximumVersion("2.7.0").IsCompatible() {
		retentionTimeFlagName = "storage.tsdb.retention"
		if p.Spec.Retention == "" {
			retentionTimeFlagValue = defaultRetention
		}
	} else if p.Spec.Retention == "" && p.Spec.RetentionSize == "" {
		retentionTimeFlagValue = defaultRetention
	}
	
	if retentionTimeFlagValue != "" {
		promArgs = append(promArgs, monitoringv1.Argument{
			Name: retentionTimeFlagName,
			Value: retentionTimeFlagValue,
		})
	}
```